[zh-cn]: Update translation of Object (#13225)
mdn/translated-content: files/zh-cn/web/javascript/reference/global_objects/object/index.md

{{JSRef}}

**`Object`** 是 JavaScript 的一种[数据类型](/zh-CN/docs/Web/JavaScript/Data_structures) 。它用于存储各种键值集合和更复杂的实体。可以通过 `{{jsxref("Object.prototype.Object", "Object()")}}` 构造函数或者使用[对象字面量](/zh-CN/docs/Web/JavaScript/Reference/Operators/Object_initializer)的方式创建对象。

## 描述

> 在 JavaScript 中，几乎所有的[对象](/zh-CN/docs/Web/JavaScript/Data_structures#objects)都是 {{jsxref("Object")}} 的实例；一个典型的对象从 `Object.prototype` 继承属性（包括方法），尽管这些属性可能被覆盖（或者说重写）。唯一不从 `Object.prototype` 继承的对象是那些 [`null` 原型对象](#null-prototype_objects)，或者是从其他 `null` 原型对象继承而来的对象。

Author: ```suggestion
在 JavaScript 中，几乎所有的[对象](/zh-CN/docs/Web/JavaScript/Data_structures#objects)都是 {{jsxref("Object")}} 的实例；一个典型的对象从 `Object.prototype` 继承属性（包括方法），尽管这些属性可能被覆盖（或者说重写）。唯一不从 `Object.prototype` 继承的对象是那些 [`null` 原型对象](#null_原型对象)，或者是从其他 `null` 原型对象继承而来的对象。
```

> 在 JavaScript 中，几乎所有的[对象](/zh-CN/docs/Web/JavaScript/Data_structures#objects)都是 {{jsxref("Object")}} 的实例；一个典型的对象从 `Object.prototype` 继承属性（包括方法），尽管这些属性可能被覆盖（或者说重写）。唯一不从 `Object.prototype` 继承的对象是那些 [`null` 原型对象](#null-prototype_objects)，或者是从其他 `null` 原型对象继承而来的对象。

Author: ```suggestion
在 JavaScript 中，几乎所有的[对象](/zh-CN/docs/Web/JavaScript/Data_structures#objects)都是 {{jsxref("Object")}} 的实例；一个典型的对象从 `Object.prototype` 继承属性（包括方法），尽管这些属性可能被覆盖（或者说重写）。唯一不从 `Object.prototype` 继承的对象是那些 [`null` 原型对象](#null_原型对象)，或者是从其他 `null` 原型对象继承而来的对象。
```

> 通过原型链，所有的对象都能观察到 `Object.prototype` 对象的改变，除非这些改变所涉及的属性和方法沿着原型链被进一步重写。尽管有潜在的危险，但这为覆盖或扩展对象的行为提供了一个非常强大的机制。为了使其更加安全，`Object.prototype` 是核心 JavaScript 语言中唯一具有[不可变原型](/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/Object/setPrototypeOf#描述)的对象——`Object.prototype` 的原型始终为 `null` 且不可更改。

Reviewer: ```suggestion
通过原型链，**所有**对象都能观察到 `Object.prototype` 对象的改变，除非这些改变所涉及的属性和方法沿着原型链被进一步重写。尽管有潜在的危险，但这为覆盖或扩展对象的行为提供了一个非常强大的机制。为了使其更加安全，`Object.prototype` 是核心 JavaScript 语言中唯一具有[不可变原型](/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/Object/setPrototypeOf#描述)的对象——`Object.prototype` 的原型始终为 `null` 且不可更改。
```

### 对象原型属性

> 你应该避免调用任何 `Object.prototype` 方法，特别是那些不打算多态化的方法（即只有其初始行为是合理的，且无法被任何继承的对象以合理的方式重写）。所有从 `Object.prototype` 继承的对象都可以自定义一个具有相同名称但语义可能与你的预期完全不同的自有属性。此外，这些属性不会被 [`null` 原型对象](#null-prototype_objects)继承。现代 JavaScript 中用于操作对象的工具方法都是[静态的](#static_methods)。更具体地说：

Author: ```suggestion
你应该避免调用任何 `Object.prototype` 方法，特别是那些不打算多态化的方法（即只有其初始行为是合理的，且无法被任何继承的对象以合理的方式重写）。所有从 `Object.prototype` 继承的对象都可以自定义一个具有相同名称但语义可能与你的预期完全不同的自有属性。此外，这些属性不会被 [`null` 原型对象](#null_原型对象)继承。现代 JavaScript 中用于操作对象的工具方法都是[静态的](#静态方法)。更具体地说：
```

- [`__proto__`](/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/Object/proto) 属性已被弃用，不应该再使用。请使用静态方法 {{jsxref("Object.getPrototypeOf()")}} 和 {{jsxref("Object.setPrototypeOf()")}} 作为替代。
- [`propertyIsEnumerable()`](/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/Object/propertyIsEnumerable) 和 [`hasOwnProperty()`](/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/Object/hasOwnProperty) 方法可以分别用静态方法 {{jsxref("Object.getOwnPropertyDescriptor()")}} 和 {{jsxref("Object.hasOwn()")}} 替换。
- 如果你正在检查一个构造函数的 `prototype` 属性，通常可以用 [`instanceof`](/zh-CN/docs/Web/JavaScript/Reference/Operators/instanceof) 代替 [`isPrototypeOf()`](/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/Object/isPrototypeOf) 方法。

> 如果不存在语义上等价的静态方法，或者你真的想使用 `Object.prototype` 方法，你应该通过 [`call()`](/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/Function/call) 直接在目标对象上调用 `Object.prototype` 方法，以防止目标对象上原有方法被覆盖，从而产生意外的结果。

Author: ```suggestion
如果不存在语义上等价的静态方法，或者你真的想使用 `Object.prototype` 方法，你应该通过 [`call()`](/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/Function/call) 直接在目标对象上调用 `Object.prototype` 方法，以防止因目标对象上原有方法被覆盖而产生意外的结果。
```

> 如果不存在语义上等价的静态方法，或者你真的想使用 `Object.prototype` 方法，你应该通过 [`call()`](/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/Function/call) 直接在目标对象上调用 `Object.prototype` 方法，以防止目标对象上原有方法被覆盖，从而产生意外的结果。

Author: ```suggestion
如果不存在语义上等价的静态方法，或者你真的想使用 `Object.prototype` 方法，你应该通过 [`call()`](/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/Function/call) 直接在目标对象上调用 `Object.prototype` 方法，以防止因目标对象上原有方法被覆盖而产生意外的结果。
```

```js
const obj = {
  foo: 1,
  // 如果可能的话，你不应该在自己的对象上定义这样的方法，

>   // 但是如果你接受外部输入，可能无法防止这种情况发生接收的对象上

Reviewer: ```suggestion
  // 但是如果你从外部输入接收对象，可能无法防止这种情况的发生
```

  propertyIsEnumerable() {
    return false;
  },
};

obj.propertyIsEnumerable("foo"); // false；预期外的结果
Object.prototype.propertyIsEnumerable.call(obj, "foo"); // true；预期的结果
```

### 从对象中删除属性

> 一个对象本身没有任何方法可以（像 {{jsxref("Map.prototype.delete", "Map.prototype.delete()")}} 一样）删除自己的属性。要删除一个对象的属性，必须使用 [delete 操作符](/zh-CN/docs/Web/JavaScript/Reference/Operators/delete)。

Reviewer: ```suggestion
一个对象本身没有任何方法可以（像 {{jsxref("Map.prototype.delete", "Map.prototype.delete()")}} 一样）删除自己的属性。要删除一个对象的属性，必须使用 [delete 运算符](/zh-CN/docs/Web/JavaScript/Reference/Operators/delete)。
```

### null 原型对象

几乎所有的 JavaScript 对象最终都继承自 `Object.prototype`（参见[继承和原型链](/zh-CN/docs/Web/JavaScript/Inheritance_and_the_prototype_chain)）。然而，你可以使用 [`Object.create(null)`](/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/Object/create) 或定义了 `__proto__: null` 的[对象字面量语法](/zh-CN/docs/Web/JavaScript/Reference/Operators/Object_initializer)（注意：对象字面量中的 `__proto__` 键不同于已弃用的 [`Object.prototype.__proto__`](/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/Object/proto) 属性）来创建 `null` 原型对象。你还可以通过调用 [`Object.setPrototypeOf(obj, null)`](/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/Object/setPrototypeOf) 将现有对象的原型更改为 `null`。

```js
const obj = Object.create(null);
const obj2 = { __proto__: null };
```

> `null` 原型对象可能会有一些预期外的行为表现，因为它不会从 `Object.prototype` 继承任何对象方法。这在调试时尤其需要注意，因为常见的对象属性转换/检测实用方法可能会生成错误或丢失信息（特别是在使用了忽略错误的静默错误捕获机制的情况下）。

Reviewer: ```suggestion
`null` 原型对象可能会有一些预期外的行为表现，因为它不会从 `Object.prototype` 继承任何对象方法。这在调试时尤其需要注意，因为常见的对象属性转换/检测实用方法可能会产生错误或丢失信息（特别是在使用了忽略错误的静默错误捕获机制的情况下）。
```

例如，[`Object.prototype.toString()`](/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/Object/toString) 方法的缺失通常会使得调试变得困难：

```js
const normalObj = {}; // 创建一个普通对象
const nullProtoObj = Object.create(null); // 创建一个 "null" 原型对象

console.log(`normalObj 是：${normalObj}`); // 显示 "normalObj 是：[object Object]"
console.log(`nullProtoObj 是：${nullProtoObj}`); // 抛出错误：Cannot convert object to primitive value

alert(normalObj); // 显示 [object Object]
alert(nullProtoObj); // 抛出错误：Cannot convert object to primitive value
```

其他方法也会失败。

```js
normalObj.valueOf(); // 显示 {}
nullProtoObj.valueOf(); // 抛出错误：nullProtoObj.valueOf is not a function

normalObj.hasOwnProperty("p"); // 显示 "true"
nullProtoObj.hasOwnProperty("p"); // 抛出错误：nullProtoObj.hasOwnProperty is not a function

normalObj.constructor; // 显示 "Object() { [native code] }"
nullProtoObj.constructor; // 显示 "undefined"
```

> 我们可以通过为 `null` 原型对象分配属性的方式将原来的 `toString` 方法添加回去：

Reviewer: ```suggestion
我们可以通过为 null 原型对象分配属性的方式将 `toString` 方法添加回去：
```

  - : 返回绑定在指定属性上的 getter 函数。
- [`Object.prototype.__lookupSetter__()`](/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/Object/__lookupSetter__) {{Deprecated_Inline}}
  - : 返回绑定在指定属性上的 setter 函数。
- {{jsxref("Object.prototype.hasOwnProperty()")}}
  - : 返回一个布尔值，用于表示一个对象自身是否包含指定的属性，该方法并不会查找原型链上继承来的属性。
- {{jsxref("Object.prototype.isPrototypeOf()")}}
  - : 返回一个布尔值，用于表示该方法所调用的对象是否在指定对象的原型链中。
- {{jsxref("Object.prototype.propertyIsEnumerable()")}}
  - : 返回一个布尔值，指示指定属性是否是对象的[可枚举自有属性](/zh-CN/docs/Web/JavaScript/Enumerability_and_ownership_of_properties)。
- {{jsxref("Object.prototype.toLocaleString()")}}
  - : 调用 {{jsxref("Object/toString", "toString()")}} 方法。
- {{jsxref("Object.prototype.toString()")}}
  - : 返回一个代表该对象的字符串。
- {{jsxref("Object.prototype.valueOf()")}}
  - : 返回指定对象的原始值。

## 示例

### 构造空对象

下面的例子将一个空的 `Object` 对象存到 `o` 中：

```js
const o1 = new Object();
const o2 = new Object(undefined);
const o3 = new Object(null);
```

### 使用 `Object` 生成布尔对象

下面的例子将 {{jsxref("Boolean")}} 对象存到 `o` 中：

```js
// 等价于 const o = new Boolean(true);
const o = new Object(true);
```

```js
// 等价于 const o = new Boolean(false);
const o = new Object(Boolean());
```

> ### Object prototypes

Author: 是否要翻译呢？

Reviewer: 可以不翻译（也可以翻译）：

```suggestion
### Object prototype
```

当我们要修改现有的 `Object.prototype` 方法时，请你考虑一下在现有逻辑之前或者之后通过包装扩展代码的方式来注入代码。例如，以下（未经测试的）代码将会在执行内部逻辑或者是其他扩展之前，有条件地执行一段自定义的逻辑。

在使用钩子修改原型时，通过在函数上调用 `apply()` 方法并传递 `this` 和参数（即调用状态），将其传递给当前行为。这种模式可以用于任何原型，例如 `Node.prototype`、`Function.prototype` 等。

```js
const current = Object.prototype.valueOf;

// 由于我的属性“-prop-value”是横跨多个原型链的，并且不总是在同一个原型链上，
// 因此我想修改 Object.prototype：
Object.prototype.valueOf = function (...args) {
  if (Object.hasOwn(this, "-prop-value")) {
    return this["-prop-value"];
  } else {
    // 这似乎不是我的对象，因此让我们尽可能实现默认行为。
    // 在某些其他语言中，apply 的行为类似于 "super"。
    // 即使 valueOf() 不需要参数，但其他的方法可能需要参数。
    return current.apply(this, args);
  }
};
```

> > **警告**：修改任何内置构造函数的 `prototype` 属性被认为是一种不好的做法，可能会影响向前兼容性。

Reviewer: 这是一种特殊的格式，冒号必须在加粗样式内

```suggestion
> **警告：** 修改任何内置构造函数的 `prototype` 属性被认为是一种不好的做法，可能会影响向前兼容性。
```

  - : 指向实例对象在实例化时使用的原型对象。
- {{jsxref("Object.prototype.constructor")}}
  - : 创建该实例对象的构造函数。对于普通的 `Object` 实例，初始值为 {{jsxref("Object/Object", "Object")}} 构造函数。其它构造函数的实例都会从它们各自的 `Constructor.prototype` 对象中继承 `constructor` 属性。

## 实例方法

- [`Object.prototype.__defineGetter__()`](/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/Object/__defineGetter__) {{Deprecated_Inline}}
  - : 将一个属性与一个函数相关联，当该属性被访问时，执行该函数，并且返回函数的返回值。
- [`Object.prototype.__defineSetter__()`](/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/Object/__defineSetter__) {{Deprecated_Inline}}
  - : 将一个属性与一个函数相关联，当该属性被设置时，执行该函数，执行该函数去修改某个属性。
- [`Object.prototype.__lookupGetter__()`](/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/Object/__lookupGetter__) {{Deprecated_Inline}}
  - : 返回绑定在指定属性上的 getter 函数。
- [`Object.prototype.__lookupSetter__()`](/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/Object/__lookupSetter__) {{Deprecated_Inline}}
  - : 返回绑定在指定属性上的 setter 函数。
- {{jsxref("Object.prototype.hasOwnProperty()")}}
  - : 返回一个布尔值，用于表示一个对象自身是否包含指定的属性，该方法并不会查找原型链上继承来的属性。
- {{jsxref("Object.prototype.isPrototypeOf()")}}
  - : 返回一个布尔值，用于表示该方法所调用的对象是否在指定对象的原型链中。
- {{jsxref("Object.prototype.propertyIsEnumerable()")}}

>   - : 返回一个布尔值，指示指定属性是否是对象的[可枚举自有属性](/en-US/docs/Web/JavaScript/Enumerability_and_ownership_of_properties)。

Author: ```suggestion
  - : 返回一个布尔值，指示指定属性是否是对象的[可枚举自有属性](/zh-CN/docs/Web/JavaScript/Enumerability_and_ownership_of_properties)。
```

- {{jsxref("Object.prototype.toLocaleString()")}}
  - : 调用 {{jsxref("Object/toString", "toString()")}} 方法。
- {{jsxref("Object.prototype.toString()")}}
  - : 返回一个代表该对象的字符串。
- {{jsxref("Object.prototype.valueOf()")}}
  - : 返回指定对象的原始值。

## 示例

### 构造空对象

> 下面的例子将一个空的 `Object` 对象存到 `o` 中：

Author: 此处原文存在问题

```js
const o1 = new Object();
const o2 = new Object(undefined);
const o3 = new Object(null);
```

### 使用 `Object` 生成布尔对象

下面的例子将 {{jsxref("Boolean")}} 对象存到 `o` 中：

```js
// 等价于 const o = new Boolean(true);
const o = new Object(true);
```

```js
// 等价于 const o = new Boolean(false);
const o = new Object(Boolean());
```

### Object prototypes

当我们要修改现有的 `Object.prototype` 方法时，请你考虑一下在现有逻辑之前或者之后通过包装扩展代码的方式来注入代码。例如，以下（未经测试的）代码将会在执行内部逻辑或者是其他扩展之前，有条件地执行一段自定义的逻辑。

在使用钩子修改原型时，通过在函数上调用 `apply()` 方法并传递 `this` 和参数（即调用状态），将其传递给当前行为。这种模式可以用于任何原型，例如 `Node.prototype`、`Function.prototype` 等。

```js
const current = Object.prototype.valueOf;

> // Since my property "-prop-value" is cross-cutting and isn't always
> // on the same prototype chain, I want to modify Object.prototype:

Author: ```suggestion
```

> // 由于我的属性“-prop-value”是横跨多个原型链的，并且不总是在同一个原型链上，因此我想修改 Object.prototype：

Author: ```suggestion
// 由于我的属性“-prop-value”是横跨多个原型链的，并且不总是在同一个原型链上，
// 因此我想修改 Object.prototype：
```

Object.prototype.valueOf = function (...args) {
  if (Object.hasOwn(this, "-prop-value")) {
    return this["-prop-value"];
  } else {

>     // It doesn't look like one of my objects, so let's fall back on
>     // the default behavior by reproducing the current behavior as best we can.
>     // The apply behaves like "super" in some other languages.
>     // Even though valueOf() doesn't take arguments, some other hook may.

Author: ```suggestion
```

>     // 这似乎不是我的对象，因此让我们尽可能地重现当前的行为，以回到默认行为。
>     // 在某些其他语言中，apply 的行为类似于 "super"。
>     // 即使 valueOf() 不需要参数，但其他钩子可能需要参数。

Author: ```suggestion
    // 这似乎不是我的对象，因此让我们尽可能实现默认行为。
    // 在某些其他语言中，apply 的行为类似于 "super"。
    // 即使 valueOf() 不需要参数，但其他的方法可能需要参数。
```
